File: Persimmon.VisualStudio.TestExplorer/VSTestRunner.fs
namespace Persimmon.VisualStudio.TestExplorer

open Microsoft.VisualStudio.TestPlatform.ObjectModel
open Microsoft.VisualStudio.TestPlatform.ObjectModel.Adapter
open System
open System.Reflection
open Persimmon.Runner.Wrapper

type VSTestRunner() =
  inherit MarshalByRefObject()

  let persimmonFamilies = [
    "Persimmon.dll"
    "Persimmon.Runner.dll"
  ]

  let loadAssembly source =
    let assemblyRef = AssemblyName.GetAssemblyName(source)
    Assembly.Load(assemblyRef)

  let collectTestCase (collector: Persimmon.Runner.Wrapper.IExecutor<Persimmon.Runner.Wrapper.TestCase>) sources =
      sources
      |> Seq.filter (fun (x: string) -> persimmonFamilies |> List.exists (fun y -> x.EndsWith(y)) |> not)
      |> Seq.map loadAssembly
      |> Seq.toArray
      |> collector.Execute
      |> Seq.map VSTestCase.ofWrapperTestCase

  let runAllTestCase (runner: Persimmon.Runner.Wrapper.IExecutor<Persimmon.Runner.Wrapper.TestCase * Persimmon.Runner.Wrapper.TestResult>) sources =
      sources
      |> Seq.filter (fun (x: string) -> persimmonFamilies |> List.exists (fun y -> x.EndsWith(y)) |> not)
      |> Seq.map loadAssembly
      |> Seq.toArray
      |> runner.Execute
      |> Seq.map (fun (c, r) -> VSTestResult.ofWrapperTestResult (VSTestCase.ofWrapperTestCase c) r)

  member __.Run<'T>(typeName: string, f: Persimmon.Runner.Wrapper.IExecutor<'T> -> unit) =
#if DEBUG
    let result = NativeMethods.ShowWaitingMessageBox("Wait on VSTestRunner.Run()...")
    let currentAppDomain = AppDomain.CurrentDomain
#endif
    let assembly = Assembly.GetExecutingAssembly()
    let fullPath = assembly.Location
    let directory = IO.Path.GetDirectoryName(fullPath)
    let setup = AppDomainSetup(LoaderOptimization = System.LoaderOptimization.MultiDomain, PrivateBinPath = directory, ApplicationBase = directory, DisallowBindingRedirects = true)
    let evidence = AppDomain.CurrentDomain.Evidence
    let appDomain = AppDomain.CreateDomain("persimmon visual studio test explorer domain", evidence, setup)
    try
      let t = appDomain.CreateInstanceAndUnwrap("Persimmon.Runner.Wrapper", typeName) :?> Persimmon.Runner.Wrapper.IExecutor<'T>
      f t
    finally
      AppDomain.Unload(appDomain)

  member this.DiscoverTests(sources: string seq, sink: ITestCaseDiscoverySink) =
#if DEBUG
    let currentAppDomain = AppDomain.CurrentDomain
    let assembly = Assembly.GetExecutingAssembly()
#endif
    this.Run("Persimmon.Runner.Wrapper.TestCollector", fun collector ->
      sources
      |> collectTestCase collector
      |> Seq.iter (fun c -> sink.SendTestCase(c))
    )
  member this.RunTests(sources: string seq, handle: IFrameworkHandle) =
    this.Run("Persimmon.Runner.Wrapper.TestRunner", fun runner ->
      sources
      |> Seq.distinct
      |> runAllTestCase runner
      |> Seq.iter handle.RecordResult
    )

File: Persimmon.VisualStudio.TestExplorer2/VSTestRunner.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using Microsoft.VisualStudio.TestPlatform.ObjectModel.Adapter;

namespace Persimmon.VisualStudio.TestExplorer2
{
    public class VSTestRunner : MarshalByRefObject
    {

        private static List<string> persimmonFamilies = new List<string>
        {
            "Persimmon.dll",
            "Persimmon.Runner.dll"
        };

        private const string runnerName = "Persimmon.Runner.Wrapper";

        private VSTestCase testCase = new VSTestCase();
        private VSTestResult testResult = new VSTestResult();

        public void DiscoverTests(List<string> sources, ITestCaseDiscoverySink sink)
        {
            this.Run<Runner.Wrapper.TestCase>("Persimmon.Runner.Wrapper.TestCollector", collector =>
            {
                foreach(var testcase in this.CollectTestCase(collector, sources))
                {
                    sink.SendTestCase(testcase);
                }
            });
        }

        public void RunTests(List<string> sources, IFrameworkHandle handle)
        {
            this.Run<Tuple<Runner.Wrapper.TestCase, Runner.Wrapper.TestResult>>("Persimmon.Runner.Wrapper.TestRunner", runner =>
            {
                foreach(var result in this.RunAllTestCase(runner, sources))
                {
                    handle.RecordResult(result);
                }
            });
        }

        private void Run<T>(string typeName, Action<Runner.Wrapper.IExecutor<T>> f)
        {
            var fullPath = Assembly.GetExecutingAssembly().Location;
            var directory = System.IO.Path.GetDirectoryName(fullPath);
            var setup = new AppDomainSetup
            {
                LoaderOptimization = LoaderOptimization.MultiDomain,
                PrivateBinPath = directory,
                ApplicationBase = directory,
                DisallowBindingRedirects = true
            };
            var evidence = AppDomain.CurrentDomain.Evidence;
            var appDomain = AppDomain.CreateDomain("persimmon visual studio test explorer domain", null, setup);
            try
            {
                f(this.CreateExecutor(appDomain, typeName));
            }
            finally
            {
                AppDomain.Unload(appDomain);
            }
        }

        private dynamic CreateExecutor(AppDomain appDomain, string typeName)
        {
            appDomain.Load(runnerName);
            return appDomain.CreateInstanceAndUnwrap(runnerName, typeName);
        }

        private IEnumerable<Microsoft.VisualStudio.TestPlatform.ObjectModel.TestCase> CollectTestCase
            (Runner.Wrapper.IExecutor<Runner.Wrapper.TestCase> collector, List<string> sources)
        {
            var cs = sources
                .Where(x => !persimmonFamilies.Any(y => x.EndsWith(y)))
                .ToList();
            return collector.Execute(cs)
                .Select(testCase.Convert);
        }

        private IEnumerable<Microsoft.VisualStudio.TestPlatform.ObjectModel.TestResult> RunAllTestCase
            (Runner.Wrapper.IExecutor<Tuple<Runner.Wrapper.TestCase, Runner.Wrapper.TestResult>> runner, List<string> sources)
        {
            var cs = sources
                .Where(x => !persimmonFamilies.Any(y => x.EndsWith(y)))
                .ToList();
            return runner.Execute(cs)
                .Select(tpl => testResult.Convert(testCase.Convert(tpl.Item1), tpl.Item2));
        }
    }
}
